R/sql-clause.R

```diff
@@ -15,79 +16,124 @@ sql_select_clauses <- function(con,
     order_by = order_by,
     limit = limit
   )
-  out <- purrr::compact(out)
-  escape(unname(out), collapse = "\n", parens = FALSE, con = con)
+  sql_format_clauses(out, lvl, con)
 }
 
-sql_clause_select <- function(con, select, distinct = FALSE, top = NULL) {
+sql_clause <- function(kw, parts, sep = ",", parens = FALSE, lvl = 0) {
+  list(
+    kw = kw,
+    parts = parts,
+    sep = sep,
+    parens = parens,
+    lvl = lvl
+  )
+}
+
+sql_clause_select <- function(con, select, distinct = FALSE, top = NULL, lvl = 0) {
   assert_that(is.character(select))
   if (is_empty(select)) {
     abort("Query contains no columns")
   }
 
-  build_sql(
-    "SELECT ",
-    if (distinct) sql("DISTINCT "),
-    if (!is.null(top)) build_sql("TOP ", as.integer(top), " ", con = con),
-    escape(select, collapse = ", ", con = con),
+  clause <- build_sql(
+    sql_kw("SELECT"),
+    if (distinct) sql_kw(" DISTINCT"),
+    if (!is.null(top)) build_sql(sql_kw(" TOP "), as.integer(top), con = con),
```

Comment:
`con` is only needed for `build_sql()`. But apart from escaping `top` the call to `build_sql()` is not necessary. Integers don't really need to be escaped so I think it is safe to remove `build_sql()` and therefore the `con` argument. What do you think?

Reply:
That seems reasonable. My only concern is that we might eventually hit a SQL variant that does something particularly weird, and we end up needing to re-thread `con` all the way down to the right place. Maybe that's unlikely because the overall form of the `SELECT` query is so highly conserved?

Reply:
It feels quite unlikely that we need `con` but with SQL you never know... So, we might as well just keep it.

Reply:
You can make the call — it'll be up to you to add it back in anyway if we later discover it is necessary 😉

Reply:
Then I hope it won't be necessary 😉